### PR TITLE
[SPARK-33685][SQL] Migrate DROP VIEW command to use UnresolvedView to resolve the identifier

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

This PR introduces `UnresolvedView` in the resolution framework to resolve the identifier.

This PR then migrates `DROP VIEW` to use `UnresolvedView` to resolve the table/view identifier. This allows consistent resolution rules (temp view first, etc.) to be applied for both v1/v2 commands. More info about the consistent resolution rule proposal can be found in [JIRA](https://issues.apache.org/jira/browse/SPARK-29900) or [proposal doc](https://docs.google.com/document/d/1hvLjGA8y_W_hhilpngXVub1Ebv8RsMap986nENCFnrg/edit?usp=sharing).

### Why are the changes needed?

To use `UnresolvedView` for view resolution. Note that there is no resolution behavior change with this PR.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Updated existing tests.

### CODE REVIEW

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
cc @cloud-fan

**Comment by imback82** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -1368,7 +1368,7 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
       sql("DROP VIEW dbx.tab1")
     }
     assert(
-      e.getMessage.contains("Cannot drop a table with DROP VIEW. Please use DROP TABLE instead"))
+      e.getMessage.contains("dbx.tab1 is a table. 'DROP VIEW' expects a view."))
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:1371]: @cloud-fan do you think the message is regressing in terms of usefulness? One thing we can do is to introduce `hint: Option[String]` to `UnresolvedView` such that we can make the message as `dbx.tab1 is a table. 'DROP VIEW' expects a view. Please use DROP TABLE instead.`. WDYT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -977,6 +982,12 @@ class Analyzer(override val catalogManager: CatalogManager)
           .map(ResolvedTable(catalog.asTableCatalog, ident, _))
           .getOrElse(u)
 
+      case u @ UnresolvedView(NonSessionCatalogAndIdentifier(catalog, ident), cmd) =>
+        u.failAnalysis(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:986]: this seems like something that we should put in `CheckAnalysis`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCommandsWithIfExists.scala [hunk 3]
```diff
@@ -17,17 +17,19 @@
...
+ * resolved. If the "ifExists" flag is set to true. the plan is resolved to [[NoopCommand]],
  */
-object ResolveNoopDropTable extends Rule[LogicalPlan] {
+object ResolveCommandsWithIfExists extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
     case DropTable(u: UnresolvedTableOrView, ifExists, _) if ifExists =>
-      NoopDropTable(u.multipartIdentifier)
+      NoopCommand(u.multipartIdentifier)
+    case DropView(u: UnresolvedView, ifExists) if ifExists =>
+      NoopCommand(u.multipartIdentifier)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCommandsWithIfExists.scala:33]: shall we add `cmdName: String` to  `NoopCommand`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:1371]: SGTM. Do we need to do the same for table commands?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCommandsWithIfExists.scala (hunk 3 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCommandsWithIfExists.scala]: Thanks, added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Moved to `CheckAnalysis`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 1 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Yes. I will go thru commands that are using `UnresolvedTable` and add hints if necessary as a follow up PR.

Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: And I will revisit https://github.com/apache/spark/pull/30475#discussion_r529022520 for `UnresolvedTableOrView`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 4]
```diff
@@ -104,6 +104,15 @@ trait CheckAnalysis extends PredicateHelper {
       case u: UnresolvedTable =>
         u.failAnalysis(s"Table not found for '${u.commandName}': ${u.multipartIdentifier.quoted}")
 
+      case u @ UnresolvedView(NonSessionCatalogAndIdentifier(catalog, ident), cmd, _) =>
+        u.failAnalysis(
+          s"Cannot specify catalog `${catalog.name}` for view ${ident.quoted} " +
+            "because view support in v2 catalog has not been implemented yet. " +
+            s"$cmd expects a view.")
+
+      case u: UnresolvedView =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:113]: do we have a clear rule about when `hint` should be used in error message?

Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:113]: The rule I am thinking is if an expected type of relation is not found. For example, `UnresolvedTable` is resolved to `ResolvedView` or `UnresolvedView` is resolved to `ResolvedTable`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:113]: then probably we should make the name clearer: `relationTypeMismatchHint`

**Review by cloud-fan — APPROVED** [review]
LGTM except one comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 4 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:113]: That sounds good, thanks!

**Comment by cloud-fan** [comment]
This is definitely not related to SparkR. I'm merging it to master, thanks!

**Pull request closed without merging**